sklearn/inspection/_permutation_importance.py

```diff
@@ -4,41 +4,36 @@
 from joblib import delayed
 
 from ..metrics import check_scoring
+from ..utils import Bunch
 from ..utils import check_random_state
 from ..utils import check_array
-from ..utils import Bunch
-
-
-def _safe_column_setting(X, col_idx, values):
-    """Set column on X using `col_idx`"""
-    if hasattr(X, "iloc"):
-        X.iloc[:, col_idx] = values
-    else:
-        X[:, col_idx] = values
-
-
-def _safe_column_indexing(X, col_idx):
-    """Return column from X using `col_idx`"""
-    if hasattr(X, "iloc"):
-        return X.iloc[:, col_idx].values
-    else:
-        return X[:, col_idx]
 
 
 def _calculate_permutation_scores(estimator, X, y, col_idx, random_state,
                                   n_repeats, scorer):
     """Calculate score when `col_idx` is permuted."""
-    original_feature = _safe_column_indexing(X, col_idx).copy()
-    temp = original_feature.copy()
+    random_state = check_random_state(random_state)
 
+    # Work on a copy of X to to ensure thread-safety in case of threading based
+    # parallelism. Furthermore, making a copy is also useful when the joblib
+    # backend is 'loky' (default) or the old 'multiprocessing': in those cases,
+    # if X is large it will be automatically be backed by a readonly memory map
+    # (memmap). X.copy() on the other hand is always guaranteed to return a
+    # writable data-structure whose columns can be shuffled inplace.
+    X_permuted = X.copy()
     scores = np.zeros(n_repeats)
+    shuffling_idx = np.arange(X.shape[0])
     for n_round in range(n_repeats):
-        random_state.shuffle(temp)
-        _safe_column_setting(X, col_idx, temp)
-        feature_score = scorer(estimator, X, y)
+        random_state.shuffle(shuffling_idx)
+        if hasattr(X_permuted, "iloc"):
+            col = X_permuted.iloc[shuffling_idx, col_idx]
```

Comment:
Speaking with @jorisvandenbossche, this is the case where one should use `values`, to discard the index.

`X_permuted.iloc[:, col_idx] = X_permuted[shuffle_idx, col_idx].values

Reply:
But `.values` would also discard dtype information for `ExtendedArray` columns, no?

Reply:
No it will keep the dtype info of the `ExtendedArray`:

```python
In [6]: s = pd.Series(pd.Categorical(['a', 'b']))                                                                                                                           

In [7]: s                                                                                                                                                                   
Out[7]: 
0    a
1    b
dtype: category
Categories (2, object): [a, b]

In [8]: s.values                                                                                                                                                            
Out[8]: 
[a, b]
Categories (2, object): [a, b]
```

That's why you need to use `.to_numpy()` to explicitly try to convert it to a numpy dtype.

Reply:
Alright, good point. Feel free to submit a new PR to simplify the code then :)

Reply:
The `.values` is a bit of a historical mix. For categorical it indeed already returns the categorical array (so preserving the dtype), but eg for periods or datetime with timezone, it does not (now in practice I suppose sklearn will care less about those dtypes).

To solve this mix, in more recent versions of pandas there is `.array` to `.values` for this use case, but so that's only available for recent versions.